src/main/scala/leon/comparison/Completor.scala

```diff
@@ -0,0 +1,261 @@
+
+    // for each function of corpus, search all common tree respective to these roots
+    // (filter only the ones containing the hole)
+    val funDefAndTrees = funDefAndRoots map { p =>
+      (p._1, p._2 flatMap possibleTrees filter hasHole)
+    }
+
+    // if no function match, return None
+    if (funDefAndTrees.isEmpty)
+      return (funDef, None, None)
+
+    val suggestion = selectBestSuggestion(expr, funDefAndTrees)
+
+    suggestion match {
+      case None => (funDef, None, None)
+      case Some(pair) => (funDef, Some(pair._1), Some(fillHole(expr, pair._2)))
+    }
+  }
+
+
+
+  def possibleRootsWithHoles(funDef_corpus: FunDef, expr: Expr): List[Value] =
+    possibleRoots(funDef_corpus.body.get, expr) filter (e => Utils.hasHole(e.b))
+
+  def getBody(funDef: FunDef): Expr = funDef.body.get
+
+  def hasHole(tree: myTree[Value]): Boolean = {
+    tree.toList map (p => p.b) exists (e => e.isInstanceOf[Choose])
+  }
+
+
+  /**
+    * select the best suggestion to fill the hole i.e. the common tree having the higher score between all the common
+    * trees and all the function of corpus
+    *
+    * @param funDefAndTrees
+    * @return
+    */
+  def selectBestSuggestion(expr: Expr, funDefAndTrees: List[(FunDef, List[myTree[Value]])]):
+  Option[(FunDef, Expr)] = {
+    val funDefAndBestTree = funDefAndTrees map { p =>
+      (p._1, selectBestTreeOption(p._2, expr, getBody(p._1)))
+    }
+
+    selectBestFun(funDefAndBestTree) match {
+      case None => None
+      case Some(pair) => Some(pair._1, findPairOfTheHole(pair._2))
+    }
+  }
+
+  def selectBestTreeOption(list: List[myTree[Value]], exprA: Expr, exprB: Expr): Option[(myTree[Value], Double)] = list match {
+    case Nil => None
+    case x :: xs => Some(selectBestTree(list, exprA, exprB))
+  }
+
+  def selectBestFun(list: List[(FunDef, Option[(myTree[Value], Double)])]): Option[(FunDef, myTree[Value])] = {
+    val listWithScore = list filterNot (p => p._2.isEmpty)
+
+    listWithScore match {
+      case Nil => None
+      case x::xs => {
+        val best = listWithScore.sortBy(p => -p._2.get._2).head
+        Some(best._1, best._2.get._1)
+      }
+    }
+  }
+
+  def scoreOptionTree(tree: Option[myTree[Value]]): Double = tree match {
+    case None => 0.0
+    case Some(t) => geometricMean(t)
+  }
+
+
+  /**
+    * Find in common tree the value that paired the hole and the expression from the corpus. The hole will be filled
+    * with this hole
+    *
+    * @param tree
+    * @return
+    */
+  def findPairOfTheHole(tree: myTree[Value]): Expr =
+    (tree.toList filter (p => p.b.isInstanceOf[Choose])).head.a
+
+  /**
+    * Really really ugly function used to travers recursively the tree until finding the hole and replace it by
+    * the wished expression.
+    *
+    * see "case Choose(_) => ..."
+    *
+    * Probably a better way to perform it, using pre-existing function that I didn't know.
```

Comment:
Thx, I changed Completor.fillHole, Utils.collectExpr and Utils.collectClass and used GenTreeOps.preMap/.fold

I still search a way to get the direct children of an Expr (as I do in Utils.getChildren). Have you got an idea?